solutionbox/code_free_ml/transform_raw_data.py

```diff
@@ -0,0 +1,253 @@
+  return args
+
+
+@beam.ptransform_fn
+def shuffle(pcoll):  # pylint: disable=invalid-name
+  return (pcoll
+          | 'PairWithRandom' >> beam.Map(lambda x: (random.random(), x))
+          | 'GroupByRandom' >> beam.GroupByKey()
+          | 'DropRandom' >> beam.FlatMap(lambda (k, vs): vs))
+
+
+def prepare_image_transforms(element, features):
+  """Replace an images url with its jpeg bytes as a web safe base64 string."""
+  from PIL import Image
+  import base64
+  from tensorflow.python.lib.io import file_io as tf_file_io
+
+  for name, transform in six.iteritems(features):
+    if transform['transform'] == IMAGE_URL_TO_VEC_TRANSFORM:
+      uri = element[name]
+      try:
+        with tf_file_io.FileIO(uri, 'r') as f:
+          img = Image.open(f).convert('RGB')
+      # A variety of different calling libraries throw different exceptions here.
+      # They all correspond to an unreadable file so we treat them equivalently.
+      # pylint: disable broad-except
+      except Exception as e:
+        logging.exception('Error processing image %s: %s', uri, str(e))
+        img_error_count.inc()
+        return
+
+      # Convert to desired format and output.
+      output = six.BytesIO()
+      img.save(output, 'jpeg')
+      image_bytes = output.getvalue()
+
+      element[name] = base64.urlsafe_b64encode(image_bytes)
+
+  return element
+
+
+def preprocess(pipeline, args):
+  input_metadata = metadata_io.read_metadata(
+      os.path.join(args.analyze_output_dir, RAW_METADATA_DIR))
+
+  schema = json.loads(file_io.read_file_to_string(
+      os.path.join(args.analyze_output_dir, SCHEMA_FILE)).decode())
+  features = json.loads(file_io.read_file_to_string(
+      os.path.join(args.analyze_output_dir, FEATURES_FILE)).decode())
+
+  column_names = [col['name'] for col in schema]
+
+  exclude_outputs = None
+  if not args.target:
+    for name, transform in six.iteritems(features):
+      if transform['transform'] == TARGET_TRANSFORM:
+        target_name = name
+        column_names.remove(target_name)
+        exclude_outputs = [target_name]
+        del input_metadata.schema.column_schemas[target_name]
+        break
+
+  if args.csv_file_pattern:
+    coder = coders.CsvCoder(column_names, input_metadata.schema, delimiter=',')
+    raw_data = (
+        pipeline
+        | 'ReadCsvData' >> beam.io.ReadFromText(args.csv_file_pattern)
+        | 'ParseCsvData' >> beam.Map(coder.decode))
+  else:
+    columns = ', '.join(column_names)
+    query = 'SELECT {columns} FROM `{table}`'.format(columns=columns,
+                                                     table=args.bigquery_table)
+    raw_data = (
+        pipeline
+        | 'ReadBiqQueryData'
+        >> beam.io.Read(beam.io.BigQuerySource(query=query,
+                                               use_standard_sql=True)))
+
+  raw_data = (
+      raw_data
+      | 'PreprocessTransferredLearningTransformations'
+      >> beam.Map(prepare_image_transforms, features))
```

Comment:
done. I think the real work will be done in tft.TransformDataset() once analyze_data.py makes a saved_mode file that does the image-string to embeddings transform.